notification/src/api/notification/notification.controller.js

```diff
@@ -1,30 +1,97 @@
 const _ = require('lodash')
 const httpUtils = require('../../utils/commons')
-const ctp = require('../../utils/ctp')
 const {
-  processNotifications,
+  processNotification,
 } = require('../../handler/notification/notification.handler')
-const config = require('../../config/config')()
+const config = require('../../config/config')
 const logger = require('../../utils/logger').getLogger()
 
-const ctpClient = ctp.get(config)
+class ValidationError extends Error {
+  constructor({ stack, message, notification, isRecoverable }) {
+    super()
+    this.stack = stack
+    this.message = message
+    this.notification = JSON.stringify(notification)
+    this.isRecoverable = isRecoverable
```

Comment:
should we manage code flow with exceptions? It's not suggested most of time because errors are expensive as they collects all stack in memory (not sure in nodejs)? Is there any other way ?

Reply:
I also read that exception is more expensive, but I don't think it makes any noticeable difference here, especially when we're dealing with network calls anyway and this app is not super high load app. So I had a look at another aspect of the problem and [in this answer](https://stackoverflow.com/a/1504400) a guy wrote:

```
Personally I like throwing Exceptions for business rule validation (not so much for user input validation) because it forces the problem to be handled upstream. If my business objects returned some kind of validation result, it could be ignored by the caller. Call me a cowboy if you wish :)
```

Since these validations are not user input validations, but misconfiguration of the app or sth wrong with Adyen's side, throwing exception is ok. Validation is also ok, but I feel like when sth is misconfigured or an expected format of notification is not coming, then it's more like an exception than validation.

Reply:
Here the client is Adyen and there is no way to be handled upstream.

Also, there is no challenge on the business calls or some cases like injection in different methods needed and you are throwing an exception to handle it, which exceptions should not be used for that purpose. 

I feel here it's not that hard to check the notification payload is valid or not. Logic is similar to what we are doing on the extension module's validations.

The only place we needed to use exception to check if the payload is a valid JSON 😉  

Note: I won't block the review because of this, you might skip my comment.